### PR TITLE
Anh & Jorge Final Tasks

### LINKED ISSUE #27: Reflect Dashboard Vote in details
1. Voting button highlight isn't reflected on the details when you vote on the dashboard
2. 
recreation
- vote on the dashboard
- go to the view discussion, the buttons will not be highlighted even though your vote was recorded

### LINKED ISSUE #31: Endpoint Auth + logout
- modify EVERY Express endpoint to check if the user is authenticated before proceeding with the endpoint logic (else throw a 403 Permission Denied error and don’t let them use the endpoint)
- implement logout

### LINKED ISSUE #33: Continuously Fetch Details Page Data (votes/comments)
Prioritize other tasks this one!
- Task: add a timer so that every 5 seconds, fetchData gets called again to update the votes / comments in details page

### LINKED ISSUE #34: Bug - Fix behavior in proposal creation amount input
Just don't allow creation of proposal if the $ amount field contains a non-numerical character
- currently, if you type 123aa, it uses an amount like $1,230 or something weird

### LINKED ISSUE #39: Proposal Details Bugs
### Vote progress bar
issue with the vote progress bars, basically it doesn’t show the vote no in %voted / % unvoted
Screen Shot 2021-04-25 at 3 10 35 PM

### fix long content being covered by progress bar / vote
If the content of the details is too long (long title), the 
voting buttons or progress bars cover the content (depending on what is conditionally rendered based on their privileges)
- potential fix is to add the progress bars into the flex div that holds the other content (title, desc, link, amt etc) so that it just renders below the rest of the content

image
Screen Shot 2021-04-25 at 2 50 03 PM

### PR DESCRIPTION
**Pull request by jorgemorales21**
## Motivation

Issues #[27](https://github.com/codebase-berkeley/tgif-voting-portal/issues/27), [31](https://github.com/codebase-berkeley/tgif-voting-portal/issues/31), [33](https://github.com/codebase-berkeley/tgif-voting-portal/issues/33), [34](https://github.com/codebase-berkeley/tgif-voting-portal/issues/34), [39](https://github.com/codebase-berkeley/tgif-voting-portal/issues/39)

## Summary of Changes

- Dashboard voting buttons now reflect user's actual vote stored in database
- Added Express middleware to verify user is authenticated before accessing any endpoint
- Implemented logout (logout button logs you out and redirects to login page)
- Added timer to continuously fetch proposal details
- Added character check to proposal management's "Amount" field so that only proposals with numbers can be submitted; if other characters are there, the box highlights in red when trying to submit
- Fixed progress bars

## Testing

A whole bunch of console.logging

## Screenshot/Gif of results

Screen Shot 2021-05-02 at 12 33 10 PM

logout-gif

### CODE REVIEW

File: client/src/components/navbar/NavBar.js [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
 import {Link} from 'react-router-dom';
+import axios from 'axios';
 
 var PRIVILEGES = 'Admin';
 
 function Navbar() {
+  async function logout() {
+    try {
+      const res = await axios.get('http://localhost:8000/logout');
+      console.log("Succesfully logged out from FRONTEND")
```
Comment (ryanhzhao) [inline client/src/components/navbar/NavBar.js:11]: Remove console.log

File: client/src/components/row/Row.js [hunk 2]
```diff
@@ -23,30 +25,50 @@ function rowDeletionMode(props) {
...
+                                props.changeTextBoxValue()}
+            }>
+            
+                <div className="proposal-title">{props.title}</div>
+                <img src={noIcon} alt='Vote Status' className="vote-status"></img>
+    
+            </div>
+            
+        );
+    }
```
Comment (ryanhzhao) [inline client/src/components/row/Row.js:70]: Not critical, but you could make this condition simpler by only putting the condition in `Vote Status</img>`

File: client/src/pages/dashboard/Dashboard.js [hunk 3]
```diff
@@ -84,7 +82,9 @@ function Dashboard() {
           user_id: 1,
           proposal_id: wantedPropID
         }
-      });
+      }, );
+      console.log("successfully submitted vote");
+      console.log(voteDecision);
```
Comment (ryanhzhao) [inline client/src/pages/dashboard/Dashboard.js]: remove console.logs

File: client/src/pages/dashboard/Dashboard.js [hunk 4]
```diff
@@ -100,6 +100,41 @@ function Dashboard() {
...
+
+  function changeToNoButton() {
+    setDashboardYesButtonClassName(dashboardUnpressedYesButtonClassName);
+    setDashboardNoButtonClassName(dashboardPressedNoButtonClassName);
+  }
+
+  async function fetchUserVote() {
+    //TODO: replace user_id in params of axios request w/ actual deserialized id from cookie (instead of hardcoded 1)
+    try {
+      const res = await axios.get('http://localhost:8000/get_one_vote', {params : {user_id: 1, proposal_id: wantedPropID}});
```
Comment (ryanhzhao) [inline client/src/pages/dashboard/Dashboard.js]: during the merging, change the hardcoded user_id to be dynamic

File: client/src/pages/dashboard/Dashboard.js [hunk 5]
```diff
@@ -100,6 +100,41 @@ function Dashboard() {
...
+      const vote = (res.data[0].vote);
+      if (vote === true) {
+        changeToYesButton();
+        console.log("changed vote to yes");
+      } else if (vote === false) {
+        changeToNoButton();
+        console.log("changed vote to no");
+      } else {
```
Comment (ryanhzhao) [inline client/src/pages/dashboard/Dashboard.js]: remove console.logs

File: client/src/pages/proposalDetails/ProposalDetails.js [hunk 6]
```diff
@@ -51,7 +51,12 @@ function ProposalConditionalRender(privileges) {
     }
     
     useEffect(() => {
-      fetchVoteInfo();
+      const intervalId = setInterval(() => {  //assign interval to a variable to clear it.
+        fetchVoteInfo();
+      }, 100)
```
Comment (ryanhzhao) [inline client/src/pages/proposalDetails/ProposalDetails.js]: may want to make this time faster. its refreshing every 100ms (0.1 seconds)

File: client/src/pages/proposalDetails/ProposalDetails.js [hunk 7]
```diff
@@ -198,9 +234,14 @@ function ProposalDetails() {
   };
 
   useEffect(() => {
-    fetchCommentData();
-    fetchProposalDetails();
-  }, []);
+    const intervalId = setInterval(() => {  //assign interval to a variable to clear it.
+      fetchCommentData();
+      fetchProposalDetails();
+    }, 100)
```
Comment (ryanhzhao) [inline client/src/pages/proposalDetails/ProposalDetails.js]: Smae thing, this is quite a fast refresh rate

File: server/app.js [hunk 8]
```diff
@@ -64,6 +64,21 @@ passport.deserializeUser(async (userID, done) => {
...
+    console.log("AUTHENTICATED - ACCESSING ENDPOINT", req.path);
+    next();
+  } else if (req.path.startsWith('/auth')) {
+    console.log("NOT AUTHENTICATED BUT ACCESSING AUTH ENDPOINT - PROCEEDING TO ENDPOINT", req.path);
+    next();
+  } else {
+    console.log("ERROR 404: NOT AUTHENTICATED - CANNOT ACCESS ENDPOINT", req.path);
+    res.status(404).send();
+  }
+}
```
Comment (ryanhzhao) [inline server/app.js]: Change status to 403 (unauthorized). 404 is (page not found)
Also, remove console.logs

File: client/src/pages/proposalManagement/ProposalManagement.js [hunk 9]
```diff
@@ -79,12 +79,16 @@ function ProposalManagement() {
...
 
     const submitProposal = async () => {
-      if (textboxValueMoney !== '' && isNaN(parseFloat(textboxValueMoney))){
-        console.log("String submited for value");
+      // if (textboxValueMoney !== '' && isNaN(parseFloat(textboxValueMoney)) && /^[a-zA-Z]/.test(textboxValueMoney)){
+      //   console.log("String submited for value");
+      //   document.getElementById("moneyNewDescription").style.borderColor="#FF0000";
+      //   document.getElementById("moneyNewDescription").style.backgroundColor="#ffcccb";
+      // }
+      // else
```
Comment (Warrenwwang1) [inline client/src/pages/proposalManagement/ProposalManagement.js]: delete

File: server/app.js [hunk 10]
```diff
@@ -172,13 +179,37 @@ app.get('/getUserVotes', async (req, res) => {
...
+    res.send(query.rows);
+  } catch (error) {
+    console.log(error.stack);
+  }
+});
+
+app.get('/get_proposals_and_user_votes', async (req, res) => {
+  try {
+    const query = await db.query(
+      'SELECT * FROM proposals LEFT JOIN (SELECT * FROM votes WHERE votes.user_id=$1) as votes_subset ON votes_subset.proposal_id = proposals.id', [1],
```
Comment (ryanhzhao) [inline server/app.js]: This is beautiful.

File: server/app.js [hunk 11]
```diff
@@ -172,13 +179,37 @@ app.get('/getUserVotes', async (req, res) => {
...
+});
+
 app.get('/get_comments', async (req, res) => {
   try {
     const proposalId = req.query.proposal_id;
     const query = await db.query(
       'SELECT * FROM comments WHERE proposal_id=$1;', [proposalId],
     );
-    console.log(query.rows)
+    console.log(query.rows);
```
Comment (ryanhzhao) [inline server/app.js]: no console.log!

File: server/app.js [hunk 12]
```diff
@@ -285,6 +316,16 @@ app.get('/get_proposal_details', async (req, res) => {
   }
 });
 
+app.get('/logout', async (req, res) => {
+  try {
+    req.logout();
+    res.send();
+    console.log("logged out from back end");
```
Comment (ryanhzhao) [inline server/app.js]: no console.log!

**Review by ryanhzhao — CHANGES_REQUESTED** [review]

File: server/app.js (hunk 8 above)
Comment (Warrenwwang1) [reply server/app.js]: feel free to keep these console.log() statements too!

**Review by ryanhzhao — APPROVED** [review]

**Pull request merged**